corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
         result_map[error].append(result)
 
     return result_map
+
+
+class IterDB(object):
+    """
+    Bulk save docs in chunks.
+
+        with IterDB(db) as iter_db:
+            for doc in iter_docs(db, ids):
+                iter_db.save(doc)
+
+        iter_db.error_ids  # docs that errored
+        iter_db.saved_ids  # docs that saved correctly
+    """
+    def __init__(self, database, chunksize=100, throttle_secs=None):
+        self.db = database
+        self.chunksize = chunksize
+        self.throttle_secs = throttle_secs
+        self.saved_ids = []
+        self.deleted_ids = []
+        self.error_ids = []
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def commit(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            results = e.results
+            error_ids = {d['id'] for d in e.errors}
+            self.error_ids.extend(error_ids)
+            success_ids = [d['id'] for d in results
+                           if d['id'] not in error_ids]
+        else:
+            success_ids = [d['id'] for d in results]
+
+        if self.throttle_secs:
+            sleep(self.throttle_secs)
+        return success_ids
+
+    def commit_save(self):
+        success_ids = self.commit(self.db.bulk_save, self.to_save)
```

Comment:
I was hoping these functions don't get called unless there's something to save (I did the check in the `__exit__`, `save` and `delete` methods)  I was thinking of these as internal methods.  I've added a leading underscore to their names and made a public `commit` method which does the check and calls both: https://github.com/dimagi/commcare-hq/commit/0a4a58b1a4bf104ade36169daf3101b0352f9ace